File: roman_numerals/clojure/roman_numerals.clj
(ns xzr
  (:use clojure.test))

(defn pop-front-digit
  [num]
  (map #(mod num %) [10000 1000 100 10]))

(defn extract-front-digit
  [nums]
  (map #(int (/ %1 %2)) nums [1000 100 10 1]))

(defn digits-of
  [num]
  (->
   (pop-front-digit num)
   (extract-front-digit)))
            
(def ones ["" "I" "II" "III" "IV" "V" "VI" "VII" "VIII" "IX"])
(def tens ["" "X" "XX" "XXX" "XL" "L" "LX" "LXX" "LXXX" "XC"])
(def hundreds ["" "C" "CC" "CCC" "CD" "D" "DC" "DCC" "DCCC" "CM"])
(def thousands ["" "M" "MM" "MMM"])
(def symbol-table [thousands hundreds tens ones])

(defn map-to-symbol
  [digits]
  (map #(%1 %2) symbol-table digits))

(defn concat-symbols
  [symbols]
  (reduce #(str %1 %2) "" symbols))

(defn arabic-to-roman
  [arabic]
  (->
   (digits-of arabic)
   (map-to-symbol)
   (concat-symbols)))

(deftest test-pop-front-digit
  (is (= (pop-front-digit 1234) [1234 234 34 4]))
  (is (= (pop-front-digit 4321) [4321 321 21 1])))
  
(deftest test-extract-front-digit
  (is (= (extract-front-digit [1234 234 34 4]) [1 2 3 4]))
  (is (= (extract-front-digit [4321 321 21 1]) [4 3 2 1])))
  
(deftest test-digits-of
  (is (= (digits-of 0) [0 0 0 0]))
  (is (= (digits-of 1234) [1 2 3 4]))
  (is (= (digits-of 4321) [4 3 2 1])))

(deftest test-map-to-symbol
  (is (= (map-to-symbol [0 0 0 0]) ["" "" "" ""]))
  (is (= (map-to-symbol [1 1 1 1]) ["M" "C" "X" "I"]))
  (is (= (map-to-symbol [3 3 3 3]) ["MMM" "CCC" "XXX" "III"]))
  (is (= (map-to-symbol [0 4 4 4]) ["" "CD" "XL" "IV"]))
  (is (= (map-to-symbol [0 9 9 9]) ["" "CM" "XC" "IX"])))

(deftest test-concat-symbols
  (is (= (concat-symbols ["" "" "" ""]) ""))
  (is (= (concat-symbols ["M" "C" "X" "I"]) "MCXI"))
  (is (= (concat-symbols ["MMM" "CCC" "XXX" "III"]) "MMMCCCXXXIII"))
  (is (= (concat-symbols ["" "CD" "XL" "IV"]) "CDXLIV"))
  (is (= (concat-symbols ["" "CM" "XC" "IX"]) "CMXCIX")))

(deftest test-arabic-to-roman
  (is (= (arabic-to-roman 1981) "MCMLXXXI")))

(run-tests 'xzr)

File: roman_numerals/python/roman_numerals.py
arabics = [1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1]
romans = ["M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"]


def to_arabic(roman):
    arabic = 0
    while len(roman) != 0:
        for i in range(len(romans)):
            if roman.startswith(romans[i]):
                arabic += arabics[i]
                roman = roman[len(romans[i]):]
    return arabic


def to_roman(arabic):
    roman = ""
    while arabic > 0:
        for i in range(len(arabics)):
            while arabic >= arabics[i]:
                roman += romans[i]
                arabic -= arabics[i]
    return roman


def test_to_arabic():
    assert(to_arabic("I") == 1)
    assert(to_arabic("II") == 2)
    assert(to_arabic("III") == 3)
    assert(to_arabic("V") == 5)
    assert(to_arabic("VIII") == 8)
    assert(to_arabic("IV") == 4)
    assert(to_arabic("XXX") == 30)
    assert(to_arabic("IX") == 9)
    assert(to_arabic("XXVIII") == 28)
    assert(to_arabic("MDCL") == 1650)


def test_to_roman():
    assert(to_roman(1) == "I")
    assert(to_roman(2) == "II")
    assert(to_roman(3) == "III")
    assert(to_roman(5) == "V")
    assert(to_roman(8) == "VIII")
    assert(to_roman(4) == "IV")
    assert(to_roman(30) == "XXX")
    assert(to_roman(9) == "IX")
    assert(to_roman(28) == "XXVIII")
    assert(to_roman(1650) == "MDCL")
